client/src/components/cullingexclusionzones/CullingExclusionZones.tsx

```diff
@@ -0,0 +1,48 @@
+import { useListUnculledZonesQuery, UnculledZone } from "../../api/liberationApi";
+import { LayerGroup, LayersControl, Circle } from "react-leaflet";
+
+interface CullingExclusionCirclesProps {
+  zones: UnculledZone[];
+}
+
+const CullingExclusionCircles = (props: CullingExclusionCirclesProps) => {
+  return (
+    <>
+      <LayerGroup>
+        {props.zones.map((zone, idx) => {
+          return (
+            <Circle
+              key={idx}
```

Comment:
https://robinpokorny.medium.com/index-as-a-key-is-an-anti-pattern-e0349aece318

The index can sometimes be reliably used as a key, but I don't think this is one of those times. Nothing differentiates the first exclusion zone between two save games. If you auto load game A (open liberation where A was the most recent save), then load game B, at least the first exclusion zone probably will not be updated.

I think the only way to use indexes as keys safely is when a parent component has a unique key, which isn't the case here.

You could try doing like I did in https://github.com/dcs-liberation/dcs_liberation/commit/c5efc908dec5f854176cd4ee9dafa66323a6d777. It's a little wasteful because ephemeral entities like these will cause component redraws on every query, but there are so few of them it doesn't really matter.

Alternatively, if we add a UUID for the `Game` itself and have that be a prop of `LiberationMap`, that might be sufficient to ensure that everything on the map is redrawn when a new game loads?

Reply:
To be honest I'm not sure, I simply based myself on `AirDefenceRangeLayer.tsx` since the principle is the same. I'll see if I can do some tests to verify that there's no unexpected behavior.

Reply:
Ok I see what you mean now, I'll give your suggestions a try and see what happens.

Reply:
I'm starting to doubt if we really need to add a UUID, cause essentially we just need to redraw culling zones whenever they change. There's multiple ways for this to happen and each of them ought to trigger a redraw on the map. Because of this I believe your second suggestion will only solve part of the problem, i.e. only when a (new) game is loaded. However if you go into your settings and change the culling distance, it ought to redraw the zones too, or at least that's what happens on v5.2 which I assume it correct behavior.

I believe the following situations ought to trigger a redraw:
 
- Changes to the culling distance setting
- Loading a (new) game
- Changes to the battlefield??? like base capture?
- Any other suggestions?

Reply:
The UUID done the way I pointed to (that is, badly, by generating an ephemeral UUID every time the zones are queried) would be enough to ensure that an old zone is not redrawn. The part we're missing is handling of the `unculled_zones_updated` field in the event stream. That'll be true whenever the zones should be requeried. Forgot that we weren't already handling that. I think if you add the UUID to use as a key and also handle that field, that'll work.

If you want to avoid the UUID, you could instead create the key based on the position and the radius (just concatenating them would be enough). Either way, you do need to listen for the push events.

Feel free to update the event stream model if the boolean isn't the way you want to go about this. You could replace that with something like:

```json
"unculled_zones": {
  "updated": true,
  "radius": 10,
  "points": [
    {"lat": 20, "lng": 30}
  ]
}
```

And fully replace the existing zones whenever that `updated` field is true (and ignore it otherwise).

I don't know if that's better or worse than the current field. My point is that it's not set in stone, so whatever API makes the server/UI code cleanest is fine with me.

Reply:
I've been at it for several hours now and I'm getting closer, though I noticed unculled zones aren't the only ones that are being difficult. There are ways to mess up the threat zones as well.

I'm slowly making progress but it's due to the fact that a lot of things have been missed. Currently I got changes in 11 files and it's not over yet, though once I get it working I'll make sure to review it all.

Also, in `settings.py` I found the following code:

````
perf_culling_distance: int = bounded_int_option(
    "Culling distance (km)",
    page=MISSION_GENERATOR_PAGE,
    section=PERFORMANCE_SECTION,
    default=100,
    min=10,
    max=10000,
    causes_expensive_game_update=True,
)
perf_do_not_cull_threatening_iads: bool = boolean_option(
    "Do not cull threatening IADS",
    page=MISSION_GENERATOR_PAGE,
    section=PERFORMANCE_SECTION,
    default=True,
)
perf_do_not_cull_carrier: bool = boolean_option(
    "Do not cull carrier's surroundings",
    page=MISSION_GENERATOR_PAGE,
    section=PERFORMANCE_SECTION,
    default=True,
    causes_expensive_game_update=True,
)
````

Why does the carrier's culling option trigger an update and while the threatening iads option doesn't? Is this a mistake? Should it be both? Or none?

Reply:
It's possible. I'm not very familiar with how culling works because I don't use it.

Reply:
Ok I think I understand why. When looking at `game.py` in `compute_unculled_zones` I can't find anything referencing the iads option. My guess it the option doesn't work at this time, therefore no need to send updates...

Reply:
More progress, I have something that works more or less, a UUID is really not needed as far as I understand, the index will do. I notice a lot of the components are just not linked correctly to their slices. It's probably one of the problems left in my solution since I can change saves and display the correct culling zones, but it's not staying on the map, i.e. the culling option disables automatically. The same happens when changing the culling distance option, which is unacceptable because you want to see how the zones change while changing the value.

Either way, I believe I'm close now...

Reply:
Turns out I still had a stupid bug left, so I got it working. Now it's just a matter of filtering out the stuff that's irrelevant...

Reply:
Solution pushed! It a big one though, I can comment like I did for end-game mechanics if that makes it easier for you.

As I mentioned earlier, the threat zones are still a problem, you can show the "full" overlay, cheat-capture a point and it will only update the first time, subsequent captures don't adjust the overlay. It's probably the next thing I'll tackle...